spurious coherence error using error-chain with nonexistent type

~~~I haven't been able to minimize this, but the code is:~~~ _(edit: see below)_

```rust
error_chain! {
    foreign_links {
        Bad(nope::Thing); // nonexistent type
    }
}
```

The first error is sensical:

```
error[E0433]: failed to resolve. Use of undeclared type or module `nope`
  --> /Users/alex/.cargo/.cargo/script-cache/expr-698cc781c70734ef/expr.rs:11:47
   |
11 |     match {error_chain! { foreign_links { Bad(nope::Thing); } }} {
   |                                               ^^^^^^^^^^^ Use of undeclared type or module `nope`
```

But the knock-on errors are not:

```
error[E0119]: conflicting implementations of trait `std::convert::From<[type error]>` for type `try_main::Error`:
  --> /Users/alex/.cargo/.cargo/script-cache/expr-698cc781c70734ef/expr.rs:11:12
   |
11 |     match {error_chain! { foreign_links { Bad(nope::Thing); } }} {
   |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |            |
   |            first implementation here
   |            conflicting implementation for `try_main::Error`
   |
   = note: this error originates in a macro outside of the current crate
```

Minimized:

```rust

struct ErrorKind;
struct Error(ErrorKind);

impl From<nope::Thing> for Error {
    fn from(_: nope::Thing) -> Self {
        unimplemented!()
    }
}

impl From<ErrorKind> for Error {
    fn from(_: ErrorKind) -> Self {
        unimplemented!()
    }
}
```

```
error[E0433]: failed to resolve. Use of undeclared type or module `nope`
 --> expanded.rs:7:11
  |
7 | impl From<nope::Thing> for Error {
  |           ^^^^^^^^^^^ Use of undeclared type or module `nope`

error[E0433]: failed to resolve. Use of undeclared type or module `nope`
 --> expanded.rs:8:16
  |
8 |     fn from(_: nope::Thing) -> Self {
  |                ^^^^^^^^^^^ Use of undeclared type or module `nope`

error[E0119]: conflicting implementations of trait `std::convert::From<[type error]>` for type `Error`:
  --> expanded.rs:13:1
   |
7  | / impl From<nope::Thing> for Error {
8  | |     fn from(_: nope::Thing) -> Self {
9  | |         unimplemented!()
10 | |     }
11 | | }
   | |_- first implementation here
12 |
13 | / impl From<ErrorKind> for Error {
14 | |     fn from(_: ErrorKind) -> Self {
15 | |         unimplemented!()
16 | |     }
17 | | }
   | |_^ conflicting implementation for `Error`
```

--- a/src/librustc/ty/mod.rs
+++ b/src/librustc/ty/mod.rs
@@ -2894,6 +2894,13 @@ impl<'tcx> TyCtxt<'tcx> {
     pub fn impls_are_allowed_to_overlap(self, def_id1: DefId, def_id2: DefId)
                                         -> Option<ImplOverlapKind>
     {
+        // If either trait impl references an error, they're allowed to overlap,
+        // as one of them essentially doesn't exist.
+        if self.impl_trait_ref(def_id1).map_or(false, |tr| tr.references_error()) ||
+            self.impl_trait_ref(def_id2).map_or(false, |tr| tr.references_error()) {
+            return Some(ImplOverlapKind::Permitted);
+        }
+
         let is_legit = if self.features().overlapping_marker_traits {
             let trait1_is_empty = self.impl_trait_ref(def_id1)
                 .map_or(false, |trait_ref| {

--- a/src/test/ui/coherence/conflicting-impl-with-err.rs
+++ b/src/test/ui/coherence/conflicting-impl-with-err.rs
@@ -1,0 +1,16 @@
+struct ErrorKind;
+struct Error(ErrorKind);
+
+impl From<nope::Thing> for Error { //~ ERROR failed to resolve
+    fn from(_: nope::Thing) -> Self { //~ ERROR failed to resolve
+        unimplemented!()
+    }
+}
+
+impl From<ErrorKind> for Error {
+    fn from(_: ErrorKind) -> Self {
+        unimplemented!()
+    }
+}
+
+fn main() {}

--- a/src/test/ui/coherence/conflicting-impl-with-err.stderr
+++ b/src/test/ui/coherence/conflicting-impl-with-err.stderr
@@ -1,0 +1,15 @@
+error[E0433]: failed to resolve: use of undeclared type or module `nope`
+  --> $DIR/conflicting-impl-with-err.rs:4:11
+   |
+LL | impl From<nope::Thing> for Error {
+   |           ^^^^ use of undeclared type or module `nope`
+
+error[E0433]: failed to resolve: use of undeclared type or module `nope`
+  --> $DIR/conflicting-impl-with-err.rs:5:16
+   |
+LL |     fn from(_: nope::Thing) -> Self {
+   |                ^^^^ use of undeclared type or module `nope`
+
+error: aborting due to 2 previous errors
+
+For more information about this error, try `rustc --explain E0433`.

